### PR TITLE
feat(rule): add rule to disallow promise without done.fail

### PR DESCRIPTION
**Pull request by yaarams**
Hey,

I've added a new linter rule to prevent promises without catch(done.fail). In case of errors these tests might fail on timeouts without informative message which can be cumbersome. Please have a look and let me know what you think. (Issue #158)

Best,
Yaara

### CODE REVIEW

File: docs/rules/no-promise-without-done-fail.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+The following patterns are not warnings:
+
+```js
+describe('A suite', function(done) {
+  it('A spec', function() {
+    asyncCall().then(function() {
+      expect(true).toBe(true);
+      done();
+    }).catch(done.fail);
```
Comment (DianaSuvorova) [inline docs/rules/no-promise-without-done-fail.md]: Is  `.then(onSuccess, done.fail)`considered to be correct? if so can you add this to the doc

File: lib/rules/no-promise-without-done-fail.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+function isSemicolon (token) {
+  return (token.type === 'Punctuator' && token.value === ';')
+}
+
+function genReport (node, context) {
+  return {
+    node: node,
+    message: 'An "it" that uses an async method should handle failure (use "done.fail")',
+    fix: function (fixer) {
```
Comment (DianaSuvorova) [inline lib/rules/no-promise-without-done-fail.js]: Nice!

File: lib/rules/no-promise-without-done-fail.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  var reject = node.arguments[1]
+  return reject &&
+    reject.type === 'MemberExpression' &&
+    reject.object.type === 'Identifier' && reject.object.name === asyncParam &&
+    reject.property.type === 'Identifier' && reject.property.name === 'fail'
+}
+
+function isPromiseThenWithoutCatch (node, asyncParam) {
+  return hasIdentifier(node, 'then') &&
+    !hasIdentifier(node.parent.parent, 'then') &&
```
Comment (DianaSuvorova) [inline lib/rules/no-promise-without-done-fail.js]: Would `parent.parent` always exist for this node?

File: test/rules/no-promise-without-done-fail.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+'use strict'
+
+const rule = require('../../lib/rules/no-promise-without-done-fail')
+const RuleTester = require('eslint').RuleTester
+const ruleTester = new RuleTester()
+
+ruleTester.run('no-promise-without-done-fail', rule, {
+  valid: [
+    {
+      code: 'it("", function(done) { somethingAsync.then(function (res) { expect(res).toBe(true); done();}).catch(done.fail)})'
```
Comment (DianaSuvorova) [inline test/rules/no-promise-without-done-fail.js]: would the rule work with arrow functions 
`it('', (done) => {
})`

File: lib/rules/no-promise-without-done-fail.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    !insideReturnStatement(node) &&
```
Comment (DianaSuvorova) [inline lib/rules/no-promise-without-done-fail.js]: why do you need this condition?

**Comment by DianaSuvorova** [comment]
Hi, Thank you for your PR, this looks really good!  I left few comments/ questions.

File: lib/rules/no-promise-without-done-fail.js (hunk 5 above)
Comment (yaarams) [reply lib/rules/no-promise-without-done-fail.js]: If you have code that you need to test asynchronously, you can declare that you receive a done callback, *return a Promise*, or use the async keyword (https://jasmine.github.io/api/edge/global.html#implementationCallback)

File: lib/rules/no-promise-without-done-fail.js (hunk 3 above)
Comment (yaarams) [reply lib/rules/no-promise-without-done-fail.js]: Yes

File: test/rules/no-promise-without-done-fail.js (hunk 4 above)
Comment (yaarams) [reply test/rules/no-promise-without-done-fail.js]: Should work, I didn't limit the solution to any type of function

Comment (DianaSuvorova) [reply test/rules/no-promise-without-done-fail.js]: if it does, please add tests cases with arrow function

File: lib/rules/no-promise-without-done-fail.js (hunk 5 above)
Comment (DianaSuvorova) [reply lib/rules/no-promise-without-done-fail.js]: Looking at the tests you provided `!insideReturnStatement(node)` doesn't catch any cases. If you comment this condition out, all of the tests pass anyways. So either add a test case that is caught by this condition or remove it. Right now this condition is unnecessary.

File: test/rules/no-promise-without-done-fail.js (hunk 4 above)
Comment (yaarams) [reply test/rules/no-promise-without-done-fail.js]: @DianaSuvorova added arrow function test cases

File: docs/rules/no-promise-without-done-fail.md (hunk 1 above)
Comment (yaarams) [reply docs/rules/no-promise-without-done-fail.md]: @DianaSuvorova updated the docs

**Comment by yaarams** [comment]
@DianaSuvorova made some fixes. Check them out :)

**Comment by DianaSuvorova** [comment]
Looks good to me. Let see if other collaborators (@burabure ) have any comments

**Comment by DianaSuvorova** [comment]
ok, let's merge it! Tank you for your contribution!

**Pull request merged**